File: solutions/c/luhn/1/luhn.c
#include "luhn.h"

#include <ctype.h>
#include <string.h>

bool luhn(const char *num) {
    size_t len = strlen(num);
    if (len <= 1) return false;

    int sum = 0;
    bool do_double = false;
    int valid_found = 0;

    for (int i = len - 1; i >= 0; i--) {
        if (isdigit(num[i])) {
            int new_num = (num[i] - '0') * (do_double ? 2 : 1);
            sum += new_num > 9 ? new_num - 9 : new_num;
            do_double = !do_double;
            valid_found++;
        } else if (!isspace(num[i]))
            return false;
    }

    return valid_found > 1 && sum % 10 == 0;
}

File: solutions/python/luhn/1/luhn.py
class Luhn:
    card = ""
    invalid = False

    def __init__(self, card_num):
        self.invalid = any(not (c.isdigit() or c.isspace()) for c in card_num)
        self.card = ''.join(num for num in card_num if num.isdigit())

    def valid(self):
        if self.card == "0" or self.invalid:
            return False

        return sum(num if num <= 9 else num - 9 for i, c in enumerate(reversed(self.card)) if (num := int(c) * 2 if i % 2 else int(c)) or True) % 10 == 0

